pkg/controller/subjectpermission/subjectpermission_controller.go

```diff
@@ -171,18 +171,91 @@ func (r *ReconcileSubjectPermission) Reconcile(request reconcile.Request) (recon
 		return reconcile.Result{}, nil
 	}
 
+	// get the Namespace instance
+	ns := &corev1.Namespace{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, ns)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, it can be transitioning to the final desired state
+			// e. g. deletion or creation still in progress. Return and retry again
+			reqLogger.Info("Object not ready")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object
+		reqLogger.Info("Error Getting Namespace")
+		return reconcile.Result{}, err
+	}
+
+	// get the NamespaceList
```

Comment:
From here until line 246 appears have some duplication/overlap with a similar block of code in `namespace_controller.go`, and it's not a trivial (eg, small) amount of duplication. It might therefor be worth creating a package to contain this logic so when the logic is changed in one place it is applied in both.

Reply:
:+1:

Reply:
Might not be able to refactor much out here. 
- Lines 174-196 are just imports. 
- Lines 200-219 is specific to this subjectPermission controller (getting clusterRoleNames, then loop through *one* subjectPermission) vs the namespace controller (loop through a *list* of subjectPermissions). 
- Lines 221-243: this involves creating a rolebinding, but then also updating the status of the reconcile. Is it possible to refactor this out? (eg r.client.Create) for both namespaceReconcile and subjectPermissionReconcile